File: frontend/src/lib/castSafety.ts
/**
 * Cast-safety classifier — frontend mirror of backend ``fpulse.types.cast_safety``.
 *
 * Operates on plain native-type names (the strings the source emits to
 * the Mapping tab, e.g. ``VARCHAR(255)``, ``NUMERIC(18,2)``, ``int4``)
 * rather than full canonical FPField objects. That's enough resolution
 * to render the per-row ✓ / ⚠ / ✕ glyph instantly without a network
 * hop — the runtime always re-classifies authoritatively on execute,
 * so this is a UX hint, not a contract.
 *
 * Mirror of the backend taxonomy:
 *   safe            lossless cast (INT → BIGINT, VARCHAR → VARCHAR widened)
 *   semantic_lossy  bytes fit, business meaning narrows (JSON → STRING)
 *   lossy           value-level loss possible (BIGINT → INT, narrowing varchar)
 *   impossible      no valid cast in principle (BLOB → DATE)
 */

export type CastTier = 'safe' | 'semantic_lossy' | 'lossy' | 'impossible';

/** Coarse logical kind we can derive from a free-form native-type string. */
type Kind =
    | 'int'
    | 'decimal'
    | 'float'
    | 'string'
    | 'bool'
    | 'date'
    | 'time'
    | 'timestamp'
    | 'json'
    | 'binary'
    | 'unknown';

/**
 * Best-effort parse of a native-type token into ``{kind, length}``.
 *
 * The Mapping tab doesn't have the full canonical FPField in hand, just
 * the source's DuckDB type and the destination's DB type. We pull out
 * just enough (kind + length/precision) to compute the dominant tier.
 */
function parseNative(raw: string): { kind: Kind; length?: number; precision?: number; scale?: number } {
    const t = (raw || '').toLowerCase().trim();
    if (!t) return { kind: 'unknown' };

    // Strings — varchar(n) / char(n) / text / uuid / inet / cidr.
    if (/varchar|character\s+varying|char\b|text|string|uuid|inet|cidr/.test(t)) {
        const m = t.match(/\((\d+)\)/);
        return { kind: 'string', length: m ? parseInt(m[1], 10) : undefined };
    }

    // Decimal / numeric — numeric(p,s) or decimal.
    if (/numeric|decimal/.test(t)) {
        const m = t.match(/\((\d+)(?:\s*,\s*(\d+))?\)/);
        if (m) return { kind: 'decimal', precision: parseInt(m[1], 10), scale: m[2] ? parseInt(m[2], 10) : 0 };
        return { kind: 'decimal' };
    }

    // Floats — real / float / double.
    if (/real|float|double/.test(t)) return { kind: 'float' };

    // Integers — smallint / integer / bigint / int2 / int4 / int8 / serial.
    if (/smallint|bigint|integer\b|^int\b|int2|int4|int8|serial/.test(t)) return { kind: 'int' };

    if (/bool/.test(t)) return { kind: 'bool' };
    if (/timestamp/.test(t)) return { kind: 'timestamp' };
    if (/^date\b/.test(t)) return { kind: 'date' };
    if (/\btime\b/.test(t) && !/timestamp/.test(t)) return { kind: 'time' };
    if (/jsonb?/.test(t)) return { kind: 'json' };
    if (/bytea|binary|blob/.test(t)) return { kind: 'binary' };

    return { kind: 'unknown' };
}

/**
 * Classify a source→target cast. Returns ``null`` if either side is
 * unparsable — the UI falls back to no glyph in that case rather than
 * misleading the operator.
 */
export function classifyCastUI(
    sourceType: string | undefined,
    targetType: string | undefined,
): { tier: CastTier; reason?: string } | null {
    if (!sourceType || !targetType) return null;
    const s = parseNative(sourceType);
    const t = parseNative(targetType);
    if (s.kind === 'unknown' || t.kind === 'unknown') return null;

    // Same kind: only difference can come from params.
    if (s.kind === t.kind) {
        if (s.kind === 'string' && s.length != null && t.length != null) {
            if (t.length >= s.length) return { tier: 'safe' };
            return { tier: 'lossy', reason: `string length narrows ${s.length} → ${t.length}` };
        }
        if (s.kind === 'decimal' && s.precision != null && t.precision != null) {
            const ss = s.scale ?? 0, ts = t.scale ?? 0;
            if (t.precision >= s.precision && ts >= ss) return { tier: 'safe' };
            return { tier: 'lossy', reason: `decimal narrows ${s.precision},${ss} → ${t.precision},${ts}` };
        }
        return { tier: 'safe' };
    }

    // Numeric widening / narrowing matrix.
    if (s.kind === 'int' && t.kind === 'decimal') return { tier: 'safe' };
    if (s.kind === 'decimal' && t.kind === 'int') return { tier: 'lossy', reason: 'fractional part truncated' };
    if (s.kind === 'int' && t.kind === 'float') return { tier: 'lossy', reason: 'precision loss for integers > 2^53' };
    if (s.kind === 'float' && t.kind === 'int') return { tier: 'lossy', reason: 'fractional part truncated' };
    if (s.kind === 'float' && t.kind === 'decimal') return { tier: 'lossy', reason: 'float→decimal can drift in low-order digits' };
    if (s.kind === 'decimal' && t.kind === 'float') return { tier: 'lossy', reason: 'decimal precision narrows to float mantissa' };

    // Date / time pairings.
    if (s.kind === 'date' && t.kind === 'timestamp') return { tier: 'safe' };
    if (s.kind === 'timestamp' && t.kind === 'date') return { tier: 'lossy', reason: 'time component dropped' };
    if (s.kind === 'time' && t.kind === 'timestamp') return { tier: 'lossy', reason: 'date component must be synthesized' };
    if (s.kind === 'timestamp' && t.kind === 'time') return { tier: 'lossy', reason: 'date component dropped' };

    // → string: semantic-lossy in most cases, lossy from binary.
    if (t.kind === 'string') {
        if (s.kind === 'json') return { tier: 'semantic_lossy', reason: 'JSON serialized as string — parseability + nested addressing lost' };
        if (s.kind === 'binary') return { tier: 'lossy', reason: 'binary→string is encoding-dependent' };
        return { tier: 'semantic_lossy', reason: 'downcast to string discards type-level constraints' };
    }

    // string → typed: lossy at runtime parse.
    if (s.kind === 'string') {
        if (t.kind === 'int' || t.kind === 'float' || t.kind === 'decimal') {
            return { tier: 'lossy', reason: 'string→numeric parses at runtime; invalid values null' };
        }
        if (t.kind === 'date' || t.kind === 'time' || t.kind === 'timestamp') {
            return { tier: 'lossy', reason: 'string→temporal parses at runtime; format-dependent' };
        }
        if (t.kind === 'bool') {
            return { tier: 'lossy', reason: 'string→boolean depends on accepted truthy/falsy values' };
        }
    }

    // JSON ↔ structured-ish — current UI doesn't see STRUCT/LIST/MAP separately.
    // (TS already narrowed t.kind ≠ 'string' via the `if (t.kind === 'string')`
    // branch above; the explicit check would have been a no-op.)
    if (s.kind === 'json') return { tier: 'semantic_lossy', reason: 'engine-specific JSON→typed mapping' };

    return { tier: 'impossible', reason: `no defined cast from ${s.kind} to ${t.kind}` };
}

/** Glyph + tier metadata for badge rendering. */
export const TIER_META: Record<CastTier, { glyph: string; label: string; tone: 'ok' | 'warn' | 'error' }> = {
    safe: { glyph: '✓', label: 'Safe', tone: 'ok' },                  // ✓
    semantic_lossy: { glyph: '⚠', label: 'Semantic-lossy', tone: 'warn' }, // ⚠
    lossy: { glyph: '⚠', label: 'Lossy', tone: 'warn' },               // ⚠ (different tooltip)
    impossible: { glyph: '✕', label: 'Impossible', tone: 'error' },     // ✕
};

File: backend/fpulse/types/cast_safety.py
"""Source → sink cast classifier.

Drives the per-row ``✓ / ⚠ / ✕`` glyph on the Mapping tab and the
preflight compatibility check on the write path.

Four-level taxonomy (from reviewer 2's expanded model):

  - ``SAFE``            lossless; original value recoverable from the target.
                         INT → BIGINT, VARCHAR(50) → VARCHAR(255), DECIMAL → DECIMAL
                         (target ≥ source precision+scale).
  - ``SEMANTIC_LOSSY``  bytes fit but business meaning narrows.
                         TIMESTAMP WITH TZ → TIMESTAMP (timezone dropped).
                         JSON → STRING (parseability lost).
  - ``LOSSY``           potential value-level loss / truncation.
                         DECIMAL(18,4) → DECIMAL(10,2), BIGINT → INT,
                         VARCHAR(500) → VARCHAR(255).
  - ``IMPOSSIBLE``      no valid cast even in principle.
                         BLOB → DATE, STRUCT → INT.

Used both by the Mapping tab (glyph rendering) and by the sink-side
preflight (which raises when ``cast_policy="safe"`` and any
classification > SAFE).
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from fpulse.types.canonical import FPField, FPType


class CastSafety(Enum):
    SAFE = "safe"
    SEMANTIC_LOSSY = "semantic_lossy"
    LOSSY = "lossy"
    IMPOSSIBLE = "impossible"


@dataclass
class CastPlanElement:
    """One row of the sink's write plan, built from a ``CanonicalSchema``."""

    source_column: str
    target_column: str
    target_native_type: str         # DDL fragment ("DECIMAL(10,2)", "NVARCHAR(MAX)")
    safety: CastSafety
    reason: str | None = None       # human-readable explanation when not SAFE


# ── Classifier ──

def classify_cast(source: FPField, target: FPField) -> tuple[CastSafety, str | None]:
    """Classify a source→target cast.

    Returns ``(safety, reason)`` where ``reason`` is a one-line
    explanation when safety is not SAFE. Examines parameterized
    metadata (precision/scale/length/timezone) to detect narrowing.
    """
    # Identity → SAFE (params still need a narrowing check below).
    if source.type == target.type:
        return _classify_same_kind(source, target)

    # Numeric widening/narrowing matrix.
    if source.type == FPType.INTEGER and target.type == FPType.DECIMAL:
        return CastSafety.SAFE, None
    if source.type == FPType.DECIMAL and target.type == FPType.INTEGER:
        return CastSafety.LOSSY, "fractional part truncated"
    if source.type == FPType.INTEGER and target.type == FPType.FLOAT:
        return CastSafety.LOSSY, "precision loss for integers > 2^53"
    if source.type == FPType.FLOAT and target.type == FPType.INTEGER:
        return CastSafety.LOSSY, "fractional part truncated"
    if source.type == FPType.FLOAT and target.type == FPType.DECIMAL:
        return CastSafety.LOSSY, "float→decimal can drift in low-order digits"
    if source.type == FPType.DECIMAL and target.type == FPType.FLOAT:
        return CastSafety.LOSSY, "decimal precision narrows to float mantissa"

    # Date/time pairings.
    if source.type == FPType.DATE and target.type == FPType.TIMESTAMP:
        return CastSafety.SAFE, None
    if source.type == FPType.TIMESTAMP and target.type == FPType.DATE:
        return CastSafety.LOSSY, "time component dropped"
    if source.type == FPType.TIME and target.type == FPType.TIMESTAMP:
        return CastSafety.LOSSY, "date component must be synthesized"
    if source.type == FPType.TIMESTAMP and target.type == FPType.TIME:
        return CastSafety.LOSSY, "date component dropped"

    # Everything-to-string is structurally compatible but semantically thin.
    if target.type == FPType.STRING:
        if source.type in (FPType.JSON, FPType.STRUCT, FPType.LIST, FPType.MAP):
            return CastSafety.SEMANTIC_LOSSY, (
                "serialized as string — parseability + nested addressing lost"
            )
        if source.type == FPType.BINARY:
            return CastSafety.LOSSY, "binary→string is encoding-dependent"
        return CastSafety.SEMANTIC_LOSSY, (
            "downcast to string discards type-level constraints"
        )

    # String → typed: always lossy (parse may fail at runtime).
    if source.type == FPType.STRING:
        if target.type in (FPType.INTEGER, FPType.FLOAT, FPType.DECIMAL):
            return CastSafety.LOSSY, "string→numeric parses at runtime; invalid values null"
        if target.type in (FPType.DATE, FPType.TIME, FPType.TIMESTAMP):
            return CastSafety.LOSSY, "string→temporal parses at runtime; format-dependent"
        if target.type == FPType.BOOLEAN:
            return CastSafety.LOSSY, "string→boolean depends on accepted truthy/falsy values"

    # JSON ↔ structured (STRUCT/LIST/MAP). Compatible storage but the
    # canonical shape lookup at the receiver depends on the engine.
    if source.type == FPType.JSON and target.type in (FPType.STRUCT, FPType.LIST, FPType.MAP):
        return CastSafety.SEMANTIC_LOSSY, "engine-specific JSON→typed mapping"
    if source.type in (FPType.STRUCT, FPType.LIST, FPType.MAP) and target.type == FPType.JSON:
        return CastSafety.SEMANTIC_LOSSY, "typed structure flattens to opaque JSON"

    # UNKNOWN ↔ anything: defer to coerce-at-runtime.
    if source.type == FPType.UNKNOWN or target.type == FPType.UNKNOWN:
        return CastSafety.LOSSY, "type unresolved at planning time"

    return CastSafety.IMPOSSIBLE, f"no defined cast from {source.type.value} to {target.type.value}"


def _classify_same_kind(source: FPField, target: FPField) -> tuple[CastSafety, str | None]:
    """Same logical kind — params decide whether it's a narrowing."""
    if source.type == FPType.DECIMAL:
        sp, ss = int(source.params.get("precision", 38)), int(source.params.get("scale", 0))
        tp, ts = int(target.params.get("precision", 38)), int(target.params.get("scale", 0))
        if tp >= sp and ts >= ss:
            return CastSafety.SAFE, None
        return CastSafety.LOSSY, f"decimal narrows {sp},{ss} → {tp},{ts}"

    if source.type == FPType.STRING:
        sl = source.params.get("length")
        tl = target.params.get("length")
        if tl is None or sl is None or int(tl) >= int(sl):
            return CastSafety.SAFE, None
        return CastSafety.LOSSY, f"string length narrows {sl} → {tl}"

    if source.type == FPType.TIMESTAMP or source.type == FPType.TIME:
        # 2026-05-22 bug fix: this branch previously read params.get("timezone")
        # but every mapper (from_postgres, future from_mssql/oracle/mysql) writes
        # params["with_timezone"]. The result was a SILENT no-op — every
        # tz-bearing → tz-naked cast scored SAFE instead of SEMANTIC_LOSSY.
        # `timezone` is kept as a fallback alias so any callers that happened
        # to write the legacy key still work, but the canonical key is
        # `with_timezone`. See test_canonical_keys_contract for the pin.
        def _tz(field) -> bool:
            return bool(field.params.get("with_timezone") or field.params.get("timezone"))
        s_tz = _tz(source)
        t_tz = _tz(target)
        if s_tz and not t_tz:
            return CastSafety.SEMANTIC_LOSSY, "timezone dropped"
        if not s_tz and t_tz:
            return CastSafety.SEMANTIC_LOSSY, "timezone synthesized from local"
        return CastSafety.SAFE, None

    if source.type == FPType.INTEGER:
        # Integer "width" lives in params.bits when known.
        sb = source.params.get("bits")
        tb = target.params.get("bits")
        if sb is None or tb is None or int(tb) >= int(sb):
            return CastSafety.SAFE, None
        return CastSafety.LOSSY, f"integer narrows {sb}b → {tb}b"

    if source.type == FPType.LIST:
        # Recurse into element_type.
        s_el = source.params.get("element_type")
        t_el = target.params.get("element_type")
        if isinstance(s_el, FPField) and isinstance(t_el, FPField):
            return classify_cast(s_el, t_el)
        return CastSafety.SAFE, None

    if source.type == FPType.MAP:
        # Both key and value must be SAFE for the map to be SAFE.
        s_k, t_k = source.params.get("key_type"), target.params.get("key_type")
        s_v, t_v = source.params.get("value_type"), target.params.get("value_type")
        worst = CastSafety.SAFE
        reasons: list[str] = []
        for s_part, t_part, label in ((s_k, t_k, "key"), (s_v, t_v, "value")):
            if isinstance(s_part, FPField) and isinstance(t_part, FPField):
                part_safety, part_reason = classify_cast(s_part, t_part)
                worst = _worst(worst, part_safety)
                if part_reason:
                    reasons.append(f"{label}: {part_reason}")
        return worst, "; ".join(reasons) or None

    if source.type == FPType.STRUCT:
        # Field-by-field; missing target fields are dropped (LOSSY).
        if source.fields is None or target.fields is None:
            return CastSafety.SAFE, None
        worst = CastSafety.SAFE
        reasons: list[str] = []
        for name, s_field in source.fields.items():
            t_field = target.fields.get(name)
            if t_field is None:
                worst = _worst(worst, CastSafety.LOSSY)
                reasons.append(f"field '{name}' dropped")
                continue
            part_safety, part_reason = classify_cast(s_field, t_field)
            worst = _worst(worst, part_safety)
            if part_reason:
                reasons.append(f"{name}: {part_reason}")
        return worst, "; ".join(reasons) or None

    # Same kind, no parameterized narrowing checks needed.
    return CastSafety.SAFE, None


_SAFETY_ORDER = {
    CastSafety.SAFE: 0,
    CastSafety.SEMANTIC_LOSSY: 1,
    CastSafety.LOSSY: 2,
    CastSafety.IMPOSSIBLE: 3,
}


def _worst(a: CastSafety, b: CastSafety) -> CastSafety:
    """Return the more-severe of two safety classifications."""
    return a if _SAFETY_ORDER[a] >= _SAFETY_ORDER[b] else b
